Refactor product unit price

jolelievre opened the pull request:
<!-----------------------------------------------------------------------------
Thank you for contributing to the PrestaShop project! 

Please take the time to edit the "Answers" rows below with the necessary information.

Check out our contribution guidelines to find out how to complete it:
https://devdocs.prestashop.com/1.7/contribute/contribution-guidelines/#pull-requests
------------------------------------------------------------------------------>

| Questions         | Answers
| ----------------- | -------------------------------------------------------
| Branch?           | develop
| Description?      | In product page you can edit your `unit_price` however in the database we don't store the edited value but instead a `unit_price_ratio` which is computed this way `unit_price_ratio = price_tax_excluded / unit_price_tax_excluded` I think this was done this way historically so that it can be applied on either a price with or without taxes and get the unit price more easily<br><br>However this leads to a few problems:<br>- it cause some precision errors due to float computation (see linked issue)<br>- it is not consistent with the way impact is stored in Combination (for unit price)<br>- in this PR that aims to handle multistore https://github.com/PrestaShop/PrestaShop/pull/25464 it causes some bugs because the DB field is related to two input fields, resulting in unexpected behaviour when updating All shops (this could be fixed by complexifying the multishop management in commands, but it seems safer and more maintainable to fix this in order to keep a cealn and understandable architecture for multishop handlers)<br>- the argument of storing ratio to ease computation on FO is understandable but on the other it requires more computation in BO to interpret the value from form, so it's not valuable enough
| Type?             | refacto
| Category?         | BO
| BC breaks?        | yes
| Deprecations?     | yes
| Fixed ticket?     | Fixes #19620 and #12078
| How to test?      | In the product page try to change the unit price of a product it should correctly impact the front as it used to be even if the database is not the same anymore We should also test that when a discount is applied (wether it's cart rule, specific price or whatever) the unit price is impacted correctly<br><br>Please keep in mind that the design is not final compared to the mock-ups no need to block this for design reasons (they will be handled later) you should focus on the feature behaviour<br><br>And although it's not related to this fix I think this issue is also fixed in the new product page https://github.com/PrestaShop/PrestaShop/issues/12078
| Possible impacts? | The PrestaShop code behaves the same since both fields are still present, however modules that perform their own internal SQL queries will have to adapt since the `unit_price_ratio` column has now been replaced with `unit_price` We also need to update the upgrade scripts in the autoupgrade module so that the content of the previous column is transferred to the new one


<!-- Click the form's "Preview" button to make sure the table is functional in GitHub. Thank you! -->

<!-- Reviewable:start -->
---
This change is [<img src="https://reviewable.io/review_button.svg" height="34" align="absmiddle" alt="Reviewable"/>](https://reviewable.io/reviews/prestashop/prestashop/26762)
<!-- Reviewable:end -->

### New feature

You can now edit the unit price with tax included:
![Capture d’écran 2021-12-01 à 17 40 01](https://user-images.githubusercontent.com/13801017/144275545-b8d48ef0-1075-4897-b061-508f2ed80091.png)

### Breaking change

- the database structure has been modified, a `product.unit_price` column has been added which is now the reference for unit price instead of the `unit_price_ratio` column which is kept for backward compatibility for now
- `ProductPricesInformation` constructor has an additional parameter `$unitPriceTaxIncluded`

### Deprecation

- the `unit_price_ratio` column is deprecated (only the column)

jolelievre committed:
Update SQL structure and replace unit_price_ratio with unit_price

jolelievre committed:
Adapt Product legacy class so that unit_price is now the reference and persisted field instead of unit_price_ratio

jolelievre committed:
Update fixtures data

jolelievre committed:
Replace computation of unit_price by computation of unit_price_ratio

jolelievre committed:
Adapt price command

jolelievre committed:
Reactivate behat tests that were previously failing

Progi1984 commented on controllers/front/ProductController.php:
```diff
@@ -1130,9 +1130,10 @@ public function getTemplateVarProduct()
-        if ($product_full['unit_price_ratio'] > 0) {
-            $unitPrice = ($productSettings->include_taxes) ? $product_full['price'] : $product_full['price_tax_exc'];
-            $product_full['unit_price'] = $unitPrice / $product_full['unit_price_ratio'];
+        $product_full['unit_price_ratio'] = $product_full['unit_price'] != 0 ? ($product_full['price_tax_exc'] / $product_full['unit_price']) : 0;
```
Is it `price_tax_exc` and not `price_tax_excl` ?

jolelievre replied:
I copied the previous code which used `price_tax_exc` (see the previous code on the left) But I'm gonna check just in case

jolelievre replied:
I don't know if it was a typo at the time, but anyway `getProductProperties` does add this field:
https://github.com/PrestaShop/PrestaShop/blob/1341e8e45856da3f6922f7a3edb4cd3081045a7c/classes/Product.php#L5645

jolelievre committed:
Add PHPDoc on query result to avoid PHPStan error on unexpected fields

jolelievre committed:
Fix divisions by zero

Hlavtox commented:
@jolelievre Take a look at this PLS 🙏 https://github.com/PrestaShop/PrestaShop/issues/23206#issuecomment-783605431

jolelievre committed:
Fix bug on cart diring checkout

jolelievre commented:
Hi @Hlavtox 

I'm not sure I understood the problem, because your comment and the initial issue seem to be about different concerns (even if both are related to unit price). So, if I get it right, you think that the unit price should automatically be recomputed by javascript in the form when we perform a modification in the product price?

Hlavtox commented:
@jolelievre I think it should. Consider a following scenario:
- You create a product with a price 100 EUR per package - 100 g. You set unit price to 1000 EUR per 1 kg. Save the product.
- ... 1 month later ...
- You set product price to 110 EUR per package. Save it. Refresh page.
- Unit price stays 1000 EUR per 1 kg - which is now wrong.

I think unit price should be recalculated when changing the base price.

jolelievre commented:
I'm not sure editing the ratio is easier for the merchant It seems more straightforward to edit the price directly instead of computing it, which is why it's the value that is actually displayed and edited in the form I'm not sure merchant actually think about ratio and then apply it to compute the actual price
My guess would be that you know the unit price from your supplier, like when you buy some meat for example you have the price per kg

About the automatic update your suggestion makes sense, but the opposite vision is also valid, for example if you have prices per kg that never move but you still want to update the product price independently thus using the difference to identify your profit then you don't want the unit price to be modified
Maybe I'm confusing a bit with "wholesale price" here, I'm not against your idea actually I like it but I want to think if it could induce some problems for some people that don't see this field like we do


But anyway these are a question for the @PrestaShop/product-team :
- first is it acceptable to change the DB from `unit_price_ratio` to `unit_price` ?
- should the `Unit price` field be automatically updated when we edit the product price?
- should we actually display the ratio instead of the unit price?

jolelievre commented:
On a side not I just notice that the unit price is never displayed unless the `unity` is filled as well So I'm thinking maybe we should add a constraint to force filling them both or they are both useless

If the constraint seems too harsh then at least a help message should indicate that both fields must be filled

jolelievre commented:
Oh and last, but not least, the unit price is only editable with tax excluded, should we add a linked field with tax included? Like for the product price?

Hlavtox commented:
@jolelievre It goes both ways.

Either you buy something per kg. You decide to sell it in 100 g pieces. So you need to divide the price anyway manually and enter it there.

Or you buy something per 100 g pieces. So you need to again compute the price manually and enter it there.

You need to know the ratio in both cases I think.

jolelievre commented:
@Hlavtox Yeah I see your point, you're probably right that either way you need to compute one of the prices But I still think it's more natural for person to enter two prices than a ratio which is much more abstract and technical.

By the way the final version of the pricing tab should look like this:

![Capture d’écran 2021-11-29 à 13 09 39](https://user-images.githubusercontent.com/13801017/143865735-58414591-542d-4229-8a24-956775a5a340.png)

So you will indeed have a preview/summary but even with this editing a ratio directly doesn't feel very easy 🤔

jolelievre commented:
@Hlavtox we discussed your suggestion, we thought about another use case. Basically, there are multiple reasons why you can change your product price, let's take your example of a product that is 100g of something, the 100g part is in the name or description.

**Case 1**
I update the product price from 100 to 150 because the cost of my product went up, so the product price must be raised and the unit price as well.

**Case 2**
I update my product from 100 to 150 because I now sell a portion of 150g instead of 100g (so I also update the description), in this case my unit price remains the same.

The feature you propose assumes we are always in case 1, when you update your product price it is actually because your unit price raised but your quantity of product didn't change so both fields are connected.

The problem is that this feature would mess with case 2, and we can't know if we are in case 1 or case 2 because in the form we actually have no relation between the quantity in the product and its unity.

I anticipate your argument ^^
`Unit price ratio reflects this relationship, if we use the ratio as an input there is no more problem`

Indeed if we only rely on the ratio and display it as an input then I can raise my product price alone without modifying the ratio so the unit price will automatically be updated. It's true it would work fine for case 1, but then what about case 2. One could argue "I'm changing the quantity in my product so I update the price and I want the ratio to be raised automatically accordingly". This is actually the current behaviour in a way.


In both cases automatic computing actually brings trouble, because we can't know the "intention" behind this price modification, so we can't make automatic changes. So it should always be the responsibility of the merchant to update one or both fields. And in that spirit, this is also why the DB persistence should match the form.

If I display the unit price and I make no modifications on the field then the DB doesn't change. Currently, because we persist the ratio there can be a modification of your DB even without changing the field, I think that's actually the main issue here.

**TLDR;**
Regardless of what we choose in our DB the form should reflect this, not some interpreted value based on two different fields. So we should match the form and the DB no matter what. And if we do that, editing the price seems more accessible and understandable for a merchant.

jolelievre committed:
Handle unit price with taxes in query handler

jolelievre committed:
Update behat tests to check unit price with taxes

jolelievre committed:
Add unit price with tax included in the form

jolelievre committed:
Handle new format to provide data for unit price

jolelievre committed:
Handle new format to build price command for unit price

jolelievre committed:
Improve equality checking to avoid overriding numbers while typing

jolelievre committed:
Handle automatic recomputing of unit price with or without taxes

jolelievre committed:
Improve a little bit the display of unit price

jolelievre committed:
Fix unit price to be recomputed with discount, also cenralize its computing in Product::getProductProperties

jolelievre committed:
Check unit price tax included after duplication

jolelievre reviewed (commented):
@Hlavtox thanks for the feedback regarding the specific prices and catalog rules

I had indeed some issues, the unit price didn't take into account the discount, it's now ok both on the Product page and in the Cart preview! It was also the occasion to optimize and centralize the code a bit Now most of the computing is performed in `Product::getProductProperties` which is used everywhere

Thanks 😉

sowbiba commented on admin-dev/themes/new-theme/js/components/form/form-object-mapper.ts:
```diff
@@ -369,6 +364,30 @@ export default class FormObjectMapper {
 
+  /**
+   * We need a custom checking method for equality, we don't use strict equality on purpose because it would result
+   * into a potential infinite loop if type don't match, which can easily happen with a number value in a text input.
```
```suggestion
   * into a potential infinite loop if type doesn't match, which can easily happen with a number value in a text input.
```

sowbiba commented on classes/Product.php:
```diff
@@ -764,9 +762,9 @@ public function __construct($id_product = null, $full = false, $id_lang = null,
-            if ($this->unit_price !== null) {
-                $fields['unit_price_ratio'] = (float) $this->unit_price > 0 ? $this->price / $this->unit_price : 0;
+        if (null === $this->update_fields || (!empty($this->update_fields['price']) && !empty($this->update_fields['unit_price_ratio']) && empty($this->update_fields['unit_price']))) {
+            if ($this->unit_price_ratio !== null) {
```
why not add this if the previous line ?

jolelievre replied:
I can indeed

sowbiba commented on classes/Product.php:
```diff
@@ -5817,17 +5815,26 @@ public static function getProductProperties($id_lang, $row, Context $context = n
+        $unitPrice = 0.0;
+        if (isset($row['unit_price'])) {
+            // Unit price is supposed to be in DB and accessible in the row
+            $unitPrice = $row['unit_price'];
```
Don't we need to cast it here ?

jolelievre replied:
I'm not sure it's required but why not!

sowbiba reviewed (commented):
First pass : few comments

MatShir commented:
> I had indeed some issues, the unit price didn't take into account the discount, it's now ok both on the Product page and in the Cart preview! It was also the occasion to optimize and centralize the code a bit Now most of the computing is performed in Product::getProductProperties which is used everywhere

Is the price per unit will have the same behavior in the FO as the product price ? A cross-barred price when they have discount ? Or the price per unit will be just be updated depending on discount ?

jolelievre committed:
Address a few feedback

jolelievre commented:
@sowbiba I addressed all your feedbacks, thanks!

MatShir commented:
Nice ! GG, LGTM :)

Progi1984 commented on src/Core/Domain/Product/QueryResult/ProductPricesInformation.php:
```diff
@@ -99,6 +105,7 @@ public function __construct(
         bool $onSale,
         DecimalNumber $wholesalePrice,
         DecimalNumber $unitPrice,
+        DecimalNumber $unitPriceTaxIncluded,
```
Please document the BC Break

jolelievre replied:
Thanks

Progi1984 reviewed (commented):
@jolelievre A small feedback about a undocumented BC Break

khouloudbelguith commented:
Hi @jolelievre,

I have some issues:

**First issue**
1. Enable Experimental product page
2. Create a product with `Retail price (tax excl.)`= `Retail price (tax incl.)` = 0
3. On page product V2: set `Retail price per unit (tax excl.)` & the `Unity` and save
4. See error: the green alert `Successful update.` is displayed but the `Retail price per unit (tax excl.)` & `Retail price per unit (tax incl.)` are not saved

https://watch.screencastify.com/v/va30PjD03GzkudjWk3oS


**Second issue**
1. Enable Experimental product page
2. Edit a product V2 when `Retail price (tax excl.)`= `Retail price (tax incl.)` <> 0
3. set `Retail price per unit (tax excl.)` with negative amount & the `Unity` and save
4. See error `An unexpected error occurred. [PrestaShop\PrestaShop\Core\Domain\Product\Exception\ProductConstraintException code 100]: Invalid product unit_price. Got "-1"`
![image](https://user-images.githubusercontent.com/16067358/145235998-5aa66e5e-8d99-479e-91d2-9f4bb72e0117.png)

**Third issue**
1. Enable Experimental product page
2. Edit a product V2 when `Retail price (tax excl.)`= `Retail price (tax incl.)` <> 0
3. set `Retail price per unit (tax excl.)` with a comma for example `2,12`
4. See an error, the price is converted to `NaN`
3. set `Retail price per unit (tax inclu.)` with a comma for example `2,12`
4. See an error, the price is converted to `NaN`

I attached a screen record
https://watch.screencastify.com/v/tocVwd4B03dFKwtG6kO2

**Fourth issue**
1. Create a new currency with an Exchange rate = 10
2. Edit a product V2 when `Retail price (tax excl.)`= `Retail price (tax incl.)` <> 0
3. Set  `Retail price per unit (tax excl.)` with valid amount & the `Unity` and save
4. Go to FO > Product details > if the default currency is selected => ok
5. Go to FO > Product details > if the other currency is selected => NOK

I attached a screen record
https://watch.screencastify.com/v/c1dpr6OzkQwexSAEBIIW

**Fifth issue**
This issue is reproduced with or without PR (Product V1 & Product V2)

1. 1. Enable Experimental product page
2. Edit a product V2 when `Retail price (tax excl.)`= `Retail price (tax incl.)` <> 0
3. set `Retail price per unit (tax excl.)` & the unity with a long-expression `tttttttttttttttttttttttttttttttttttttttttttttttttttttttttttttttttttttttttttttttttttttttttttttttttttttttttttttttttttttttttttttttttttttttttttttttttttttttttttttttttttttttttttttttttttttttttttttttttttttttttttttttttttttttttttttttttttttttttttttttttttttttttttttt` 
4. Go to FO => Product details page
5. See error 
![image](https://user-images.githubusercontent.com/16067358/145243426-7cf2b5ed-0ebc-4dd4-994b-855693582228.png)

Thanks!

Hlavtox commented:
@khouloudbelguith Issue 5 can happen in any field in prestashop and there is no way to solve it, if it does not contain any spaces

If you set it to break on any letter, it could easily display like this on some displays. We dont want that. ;)

```
per 100 gr
ams
```

jolelievre committed:
Improve validation of unit price

jolelievre committed:
Add constraint for price to be positive or zero

jolelievre committed:
Object mapper only riggers change on change not keyup any more, it reduces the number of computation and prevents bug when using number formatter which triggers on change

jolelievre committed:
Comma transformer is also applied on MoneyType fields

jolelievre committed:
Apply css ellipsis on unit price div

jolelievre committed:
Apply css ellipsis on unit price div in cart display, and fix the css selector which was ignored because there is now an intermediate div so the unit price was not direct descendant any more

jolelievre committed:
Convert unit price to correct currency

jolelievre commented:
Hi @Hlavtox

actually, you can apply an ellipsis in CSS for these cases (combining overflow and text-overflow properties), you can check the modif in scss for more details.

Hi @khouloudbelguith 

I addressed issues 2 to 5, it should all work fine now. I also fixed the retail prices inputs in the product page which had the same NaN problem, along with negative values. The negative values are now mentioned with inline errors inside the form for each price input.

Regarding the first issue it's normal, you can't have a price zero and a unit price not zero, since unit price is a portion a the price, if the price is zero it is necessarily zero too. I don't know if this is mentioned in the specs but it was already the case in the legacy page, but it's less visible:

- if you test the legacy page on this branch you can indeed manage to have a price zero, and a unit price not zero But it's only thanks to the modification from this PR which can now store the unit price and not the ratio However in the new page we correctly update the unit price and set it to zero
- if you test the legacy page on develop branch (or previous version) you can also set a price zero and let the unit price unchanged and the save is successful. However, if you refresh the page you will see the the unit price has been reset to zero It's just less obvious because the page relies on ajax updates

Hlavtox commented:
@jolelievre Yeah, but no information should be hidden IMHO, still better if it overflows. 🧐

sowbiba commented on classes/Product.php:
```diff
@@ -764,10 +762,13 @@ public function __construct($id_product = null, $full = false, $id_lang = null,
-            }
+
+        // In case unit_price is not in fields we can try and fallback with unit_price_ratio
+        if (null !== $this->unit_price_ratio && empty($fields['unit_price']) && (
```
Shouldn't ```empty($fields['unit_price'])``` be the first criteria ?

jolelievre replied:
I'm not sure both make sense and I don't think the order is important here anyway

sowbiba commented on classes/Product.php:
```diff
@@ -5817,17 +5818,30 @@ public static function getProductProperties($id_lang, $row, Context $context = n
+            $unitPrice = (float) $row['unit_price'];
+        } elseif (isset($row['unit_price_ratio']) && 0 != $row['unit_price_ratio']) {
+            // In case unit_price is not defined but ratio is we recompute unit_price based on initial product price
+            $unitPrice = ($row['price_without_reduction_without_tax'] / $row['unit_price_ratio']);
```
```suggestion
            $unitPrice = $row['price_without_reduction_without_tax'] / $row['unit_price_ratio'];
```

sowbiba commented on classes/Product.php:
```diff
@@ -5817,17 +5818,30 @@ public static function getProductProperties($id_lang, $row, Context $context = n
+        $unitPrice = Tools::convertPrice($unitPrice, $currencyId);
+
+        // Compute price ratio based on initial product price and initial unit price (without taxes and without discount)
+        $row['unit_price_ratio'] = $unitPrice != 0 ? ($row['price_without_reduction_without_tax'] / $unitPrice) : 0.0;
```
```suggestion
        $row['unit_price_ratio'] = $unitPrice != 0 ? $row['price_without_reduction_without_tax'] / $unitPrice : 0.0;
```

sowbiba reviewed (commented):
Some details

jolelievre committed:
Remove useless parentheses

khouloudbelguith commented:
Hi @jolelievre,

I have some issues:
**First issue**
1. Install a shop with English language + France country
2. Change the employe lanaguge to Frensh
3. Edit a product V2 when `Retail price (tax excl.)= Retail price (tax incl.) `<> 0
4. set Retail price per unit (tax excl.) and save 
![image](https://user-images.githubusercontent.com/16067358/145808687-9a169245-7c2e-4365-9704-530b58ac0127.png)
5. Now change the tax rule applied
6. See error 
![image](https://user-images.githubusercontent.com/16067358/145808755-c5ddb0a5-cbcc-494d-b35d-ea874ffa6021.png)


**Second issue**

1. Enable Experimental product page
2. Edit a product V2 when Retail price (tax excl.)= Retail price (tax incl.) <> 0
3. set Retail price per unit (tax excl.) & the unity with a long-expression `tttttttttttttttttttttttttttttttttttttttttttttttttttttttttttttttttttttttttttttttttttttttttttttttttttttttttttttttttttttttttttttttttttttttttttttttttttttttttttttttttttttttttttttttttttttttttttttttttttttttttttttttttttttttttttttttttttttttttttttttttttttttttttttt`
4. Go to FO => Product details page => the long text is replaced by 3points `...`, is it the wanted behavior? 
![image](https://user-images.githubusercontent.com/16067358/145809193-ca0a272e-1acf-48e5-b38d-a601c944a63b.png)

Same in the shopping cart page 
![image](https://user-images.githubusercontent.com/16067358/145809276-182c4117-85a6-4479-86b5-f6acd5bf547e.png)

**Third issue**
In the FO > shopping cart, I have differences with & without PR

**With PR** 
1. If a specific price is not applied 
![image](https://user-images.githubusercontent.com/16067358/145809502-550347e5-ccb8-4476-a44a-5d9a7b88c3de.png)
**Without PR** 
![image](https://user-images.githubusercontent.com/16067358/145809618-8a1a1e4f-42b4-4055-b331-4dc4025c7843.png)
2. If a specific price is applied
**With PR** 
![image](https://user-images.githubusercontent.com/16067358/145809763-9e330f27-4942-48f0-acd0-ee8885eb8fc8.png)

**Without PR** 
![image](https://user-images.githubusercontent.com/16067358/145809819-5980cfad-503c-4c70-8071-ae520d440407.png)


**Fourth issue** 
1. Go to BO > Shop Parameters > Customer Settings > Groups page & change Price display method to tax excluded 
![image](https://user-images.githubusercontent.com/16067358/145809988-010f43f4-72c3-492c-91e2-a05e9a658baf.png)
2. Create a product with retail price = 10 Euro tax excluded & `Retail price per unit (tax excl.)` = 10 Euro 
3. Create a specific price for this product 
![image](https://user-images.githubusercontent.com/16067358/145810447-4a637e78-92d4-4a96-832e-e34b919b338d.png)

4. Go to FO > View this product
5. See error => the unit price is displayed without discount 
![image](https://user-images.githubusercontent.com/16067358/145810391-9e9ba42a-f7ec-4746-9cfd-78208034c4fe.png)


**Fifth issue**
1. Go to BO > Shop Parameters > Customer Settings > Groups page & add a discount 20% 
![image](https://user-images.githubusercontent.com/16067358/145811362-c51a7567-0dd1-4534-b1eb-3ed4087855b2.png)
2. Create a product with retail price = 10 Euro tax excluded & `Retail price per unit (tax excl.)` = 10 Euro 
3. Go to FO > Check the product
4. See error: the unit price is displayed without discount  
![image](https://user-images.githubusercontent.com/16067358/145811498-6103b4a0-652a-4e4c-b56a-8843667cccb5.png)


Thank you!

jolelievre committed:
Integrate getBigNumber in the product mapper which internally clean the input values in case commas are used

jolelievre committed:
Reapply color on cart unit price discount

jolelievre committed:
Recompute unit price tax excluded to include discount

jolelievre committed:
Use price without any reduction to compute the ratio correctly

jolelievre commented:
Hi @khouloudbelguith 

thanks for the feedback again 🙂

### First issue
It has been fixed by using our JS tools `commaTransformer` which converts number using commas `3,78` into regular float values using dot `3.78` This fixes the bug but it also changes the display of the value inside the field which is considered as a valid way to handle this feature

### Second issue
About the ellipsis indeed, I have no other way but to truncate the long text in order not to break the design. This is the strategy that @TristanLDD and @NeOMakinG advise in the BO so I figured it is acceptable in FO as well (let's remember this is an edge case that has very few chances of happening in real life). This should be approved by them or @PrestaShop/product-team 

### Third issue

About the cart rendering and the differences in color with this PR, it is indeed on purpose I stumbled upon this CSS rule:
https://github.com/PrestaShop/PrestaShop/blob/a0854ba43e3475c7ccc273feeeb4d6a46c17b958/themes/classic/_dev/css/components/cart.scss#L300

Which is from 2015-16 and was not effective anymore because the layout changed a little and the rule was too strict. So I modified it a bit to re-enable it, and the effect is that unit price is displayed differently from the price (smaller, passing and different color). This makes sense to me but that also needs to be validated by @NeOMakinG @TristanLDD and/or @PrestaShop/product-team 

However, you are right when a discount is applied, the color of the unit price should also be affected, this has now been fixed and it uses the same discount color as the product price.

### Fourth issue
The unit price discount was not applied on tax excluded value, it has now been fixed.

### Fifth issue
Group discount was not correctly applied to unit price, this has now been fixed

jolelievre committed:
Unit price in cart is bllue but lighter

jolelievre committed:
Final display of unit price

jolelievre commented:
Hi @khouloudbelguith 

we validated with @MatShir @NeOMakinG and @prestascott (the whole gang) how the unit price was supposed to be displayed. Here is the final decision:
- same display for product page and cart
- the color is always gray (even when there is a promo)
- the text is smaller and lighter
- no parentheses in both views

![Capture d’écran 2021-12-14 à 11 18 08](https://user-images.githubusercontent.com/13801017/145980346-15a8b17f-2af0-4fc0-b51e-280b2b9e588a.png)
![Capture d’écran 2021-12-14 à 11 17 57](https://user-images.githubusercontent.com/13801017/145980378-cfd04528-2e53-4d85-b96e-99f425606d9c.png)

eternoendless commented on admin-dev/themes/new-theme/js/components/form/form-object-mapper.ts:
```diff
@@ -369,6 +377,30 @@ export default class FormObjectMapper {
   }
 
+  /**
+   * We need a custom checking method for equality, we don't use strict equality on purpose because it would result
+   * into a potential infinite loop if type doesn't match, which can easily happen with a number value in a text input.
+   *
+   * And we also try to see if both values have the same Number value, this avoids forcing a number input value when
+   * it's not written exactly the same way (like pending zeros)
```
These are implementation details that belong in the method's body, not in the header.

The header documentation is meant to describe what the method does (i.e. what mutation it performs if it's a command, and what information it returns if it's a query) so that you don't need to reverse engineer it, hence the importance of describing the parameters and return values beyond the types.

eternoendless commented on admin-dev/themes/new-theme/js/components/form/form-object-mapper.ts:
```diff
@@ -369,6 +377,30 @@ export default class FormObjectMapper {
+   * @param value
+   * @private
+   */
+  private hasSameValue(
+    $input: JQuery,
+    value: string | number | string[] | undefined,
+  ): boolean {
+    // eslint-disable-next-line eqeqeq
+    if (Number(numberCommaTransform(value)) == numberCommaTransform($input.val())) {
+      return true;
+    }
+
+    // eslint-disable-next-line eqeqeq
+    return value == $input.val();
+  }
```
It's unclear to me what this function is meant to do, but it's is mixing up view management (loading a field's value with jquery) with business logic. You should extract the `.val()` from and pass it as a parameter.
```suggestion
  private areEqual(
    inputValue: string,
    referenceValue: string | number | string[] | undefined,
  ): boolean {
    // eslint-disable-next-line eqeqeq
    if (Number(numberCommaTransform(referenceValue)) == numberCommaTransform(inputValue)) {
      return true;
    }

    // eslint-disable-next-line eqeqeq
    return referenceValue == inputValue;
  }
```

jolelievre replied:
Yep, good idea

sowbiba commented on classes/Product.php:
```diff
@@ -5814,17 +5815,46 @@ public static function getProductProperties($id_lang, $row, Context $context = n
+            true,
+            false
+        );
+        $row['unit_price_ratio'] = $baseUnitPrice != 0 ? $baseProductPrice / $baseUnitPrice : 0.0;
```
if baseUnitPrice is equal to 0, the baseProductPrice get is useless

jolelievre replied:
Thanks

jolelievre committed:
Address feedback from review

jolelievre commented:
Thanks @sowbiba and @eternoendless I just addressed your feedback

eternoendless commented on classes/Product.php:
```diff
@@ -764,10 +762,13 @@ public function __construct($id_product = null, $full = false, $id_lang = null,
+            null === $this->update_fields ||
+            (!empty($this->update_fields['price']) && !empty($this->update_fields['unit_price_ratio']) && empty($this->update_fields['unit_price']))
+        )) {
+            $fields['unit_price'] = (float) $this->unit_price_ratio > 0 ? $this->price / $this->unit_price_ratio : 0;
```
Shouldn't this be done using decimal to avoid loss of precision?

jolelievre replied:
Maybe but honestly I kept this code to avoid changing too many things I see no reason why this could happen

PierreRambaud replied:
I'm not sure we have to put the decimal library everywhere, maybe in another pull request.
Because it makes more sense to use decimal for `unit_price_tax_excluded` or `unit_price_tax_included`.
I want to trust @jolelievre when it concerns prices :eyes:

jolelievre replied:
Actually, this code is now useless so I'm gonna clean it because it's now useless

jolelievre commented:
As discussed with @PierreRambaud I'm gonna modify this PR to reduce the BC break, we are going to keep the `unit_price_ratio` in the DB for now, but it will be updated automatically by the Product class after each update

jolelievre committed:
We keep unit_price_ratio in the DB for backward compatibility but it is automatically updated based on unit_price

jolelievre committed:
Add deprecation on DB field

jolelievre committed:
Improve deprecation comment

khouloudbelguith commented:
Hi @jolelievre,

I have an issue with a product with combinations:
Steps to reproduce the issue:
1. Go to BO > Shop Parameters > Products Settings > `Quantity discounts based on` Products 
![image](https://user-images.githubusercontent.com/16067358/146761582-cdaf1b6f-9c8a-48c1-b30d-62b259f49888.png)
2. Create a product with 2 combinations with V2 `retail price = 100 Euro tax excluded` & `Retail price per unit (tax excl.) = 100 Euro` 
![image](https://user-images.githubusercontent.com/16067358/146762158-23ce3220-b959-42eb-8c0c-90f30f40732e.png)
3. Create a specific price from quantity = 2 
![image](https://user-images.githubusercontent.com/16067358/146762891-056c43d5-2b67-4d01-9846-10209416dc79.png)
4. Go to FO > add the first combination to cart with quantity =1 / add the second combinations with quantity = 1 
![image](https://user-images.githubusercontent.com/16067358/146763362-03787368-5a3f-41d0-a36b-7824a4155ce8.png)
5. Go to FO > Shopping cart page
6. See error the  unit price is displayed without discount 
![image](https://user-images.githubusercontent.com/16067358/146763337-8a088922-d301-446f-afb1-b22adde00cc5.png)

It should be displayed with a discount because we have 2 combinations of the same product in the cart

Thanks!

jolelievre committed:
Apply cart discounts to cart unit price

kpodemski commented on classes/Cart.php:
```diff
@@ -970,6 +970,10 @@ protected function applyProductCalculations($row, $shopContext, $productQuantity
 
+        // Update unit price in case cart reductions happened
+        $row['unit_price'] = $row['unit_price_tax_excluded'] = $row['unit_price_ratio'] != 0 ? $row['price_with_reduction_without_tax'] / $row['unit_price_ratio'] : 0.0;
+        $row['unit_price_tax_included'] = $row['unit_price_ratio'] != 0 ? $row['price_with_reduction'] / $row['unit_price_ratio'] : 0.0;
```
@jolelievre could you test it with the gift automatically added to the cart that has the unit price?

PierreRambaud commented:
I approve your PR @jolelievre but I don't know if it's possible to add test as @kpodemski ask.
It can be done in another PR imho

jolelievre commented:
There is no automated test on this FO part, and I'm not sure I understand exactly what you mean
@kpodemski could you give a detailed scenario that would be validated in QA as well?

kpodemski commented:
@jolelievre @PierreRambaud I've checked some scenarios with the gift (added via cart voucher) myself and everything is all right 👍

khouloudbelguith commented:
Hi @jolelievre,

I have an issue following these steps:
1. Go to BO > Shop Parameters > Products Settings > `Quantity discounts based on` Products 
![image](https://user-images.githubusercontent.com/16067358/146761582-cdaf1b6f-9c8a-48c1-b30d-62b259f49888.png)
1. Create a product with 2 combinations with `Retail price(tax excl.) = 10`  & `Retail price per unit (tax excl.)=10` 
![image](https://user-images.githubusercontent.com/16067358/147061921-e73ed18b-a7e8-4335-bbd8-f2ffb6f89206.png)
2. For each combination has an `Impact on price (tax excl.)= 10` & `Impact on price per unit (tax excl.) =10` 
![image](https://user-images.githubusercontent.com/16067358/147062393-fcd9f62f-fab1-4a55-845c-a236903d6590.png)
3. Create a first specific price on all combinations `Starting at 2` and `Apply a discount of 20 Euro tax included`
4. Create a second specific price on the Second combination `Starting at 3` and `Apply a discount of 50%`
![image](https://user-images.githubusercontent.com/16067358/147063444-817b2aa4-83f5-4cc9-a149-3feecf66807f.png)
5. Go to FO > Add first combination with quantity = 1 & second combination with quantity = 1 
![image](https://user-images.githubusercontent.com/16067358/147063564-6556f3cb-9e57-4885-9aa1-4e73441b27da.png)
6. In the FO > Shopping carts page 
![image](https://user-images.githubusercontent.com/16067358/147064731-e71bba1d-01bc-47b1-9b63-994c9f2b251c.png)
7. Increase the quantity of the seond combination 
8. See error
![image](https://user-images.githubusercontent.com/16067358/147065086-f09b5390-3599-494d-8d4b-2425ec34cff0.png)

PS: if we increase the quantity to 3 => ok 
![image](https://user-images.githubusercontent.com/16067358/147067906-af293c51-1f7b-4763-8937-b685f6843e31.png)


**Second issue**
1. Create a product with `Retail price (tax excl.)=15.706723` 4 numbers after the comma & `Display retail price per unit tax excl = 100` & No tax rule 
![image](https://user-images.githubusercontent.com/16067358/147076261-b2c3751e-cac3-48f4-bded-038ba83c6f96.png)
2. Go to FO > View product
3. See error 
![image](https://user-images.githubusercontent.com/16067358/147076419-028f574d-12c9-47f8-a210-c158f8ce5602.png)

I created a separate issue about tax rule: https://github.com/PrestaShop/PrestaShop/issues/27054 not related to this PR

Thanks!

jolelievre commented:
Hi @khouloudbelguith 

regarding the first issue, the problem doesn't come from the unit price but rather the way specific price is applied in the cart. If you look at the product line its price is 24€ for two products, so it's indeed 12€ per product. The unit price is consistent based on the line total price, it's the cart computing of the line which is not. The product should have a -20€ discount in this case, so its line value should be 2 * 4 = 8€

Other cases seem to happen fine
![Capture d’écran 2021-12-29 à 12 06 50](https://user-images.githubusercontent.com/13801017/147657375-8d45c825-0332-47be-9b77-581ff2a57e8f.png)

![Capture d’écran 2021-12-29 à 12 08 55](https://user-images.githubusercontent.com/13801017/147657409-f59e8416-cc93-4bab-8847-75ab40e33fce.png)

but this one is buggy:
![Capture d’écran 2021-12-29 à 12 03 36](https://user-images.githubusercontent.com/13801017/147657359-60ceaa04-e0d8-46a9-9b6f-016bc9a01646.png)

This bug is not related to this PR so it should be handled in a dedicated issue. Besides I checked on develop, we have the exact same bug, so it's not a regression introduced by this PR. Therefore the first issue should not block this PR.

I'm currently checking for the second issue though.

jolelievre committed:
Save prices before rounding to avoid rouding issues on unit prices, little refacto for cleaner code

jolelievre commented:
Ok, the second issue has been fixed. I hope it's the last one ^^

khouloudbelguith commented:
Hi @jolelievre,

Sorry, but I still have some issues:

**First issue**

1. Enable Experimental product page
2. Go to BO > Catalog > Products page
3. Add product V2 > in the pricing tab > change Tax rule to 10%
4. set the `Retail price (tax incl.)= 500` & `Retail price per unit (tax incl.) =500`
5. Click outside
6. See error => it will be converted to `500.000001`

https://watch.screencastify.com/v/gZHB5BzmLoOIP2GClcHE


**Second issue**

1. Enable Experimental product page
2. Go to BO > Catalog > Products page
3. Edit Product V2 >  in the pricing tab > change the `Retail price (tax incl.)`
4. Don't click outside and click on save directly
5. See error the `Retail price` is not changed
6. Same issue if you change the `Retail price per unit (tax incl.)`
7. Don't click outside and click on save directly
8. See error the `Retail price per unit` is not changed

https://watch.screencastify.com/v/PduZFUpuoeDVcJ1SstoS

Thanks!

jolelievre committed:
Debounce is not needed anymore since only change event is watched this fixes a bug when form is submitted before change is applied

jolelievre commented:
Hi @khouloudbelguith 

thanks for the feedback.

### Second issue
I just fixed the second issue, and changes should also be applied faster now, there was a debounce delay of 350ms which caused this bug (the debounce is not needed anymore because we don't watch every `keyup` even anymore). So this one is fixed.

### First issue
Regarding the first issue, actually it's a legacy bug that is also present on the legacy page. You can test it with the same values you will have the same problem on tax included value. You don't see it right away but when you refresh the page you will see the same incorrect value.

The problem is bound to our way of storing price values in DB, since we only save the price tax excluded we have this one digit error when applying the tax. To fix this completely we would need to change our DB structure to store both values (and maybe indicate which one is used as reference). But it's not the purpose of this PR, it's not even in the roadmap as far as I know.

So since the bug was already here it's not a regression introduced by this PR, therefore it shouldn't block it (the bug on unit price is exactly for the same reason, it was not visible before since the tax included input didn't exist).

jolelievre commented:
@khouloudbelguith this one is ready for QA

If the submit bug was the last one everything should be fine now

khouloudbelguith commented:
Hi @jolelievre,

It is ok :heavy_check_mark: 
https://github.com/PrestaShop/PrestaShop/issues/19620 fixed: https://watch.screencastify.com/v/rNDGm7iA8ybrpjAax8Yr
https://github.com/PrestaShop/PrestaShop/issues/12078 fixed: https://watch.screencastify.com/v/wc3sbpCrYhRfTxkSBrls

I created issues previously reported which are not the scope of this PR:
https://github.com/PrestaShop/PrestaShop/issues/27229
https://github.com/PrestaShop/PrestaShop/issues/27230

Thank a lot for your help!

jolelievre commented:
Thank you @khouloudbelguith
